### PR TITLE
Fix completions in switch-case for type base switch for Java17 and above #1301

### PR DESCRIPTION
**Pull request by gayanper**
## What it does
Compute type completions for switch statements which are type base when the Java version is 17 or greater.

## How to test
Provided test case and unit tests.

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by gayanper** [comment]
@stephan-herrmann @trancexpress Will one of you have time to have a look at this PR ?

**Comment by stephan-herrmann** [comment]
> @stephan-herrmann @trancexpress Will one of you have time to have a look at this PR ?

I'm looking at it now.

File: org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java [hunk 1]
```diff
@@ -3989,6 +3989,8 @@ private void completionOnSingleNameReference(ASTNode astNode, ASTNode astNodePar
 					}
 					System.arraycopy(keywords, 0, keywords = new char[count][], 0, count);
 					findKeywords(this.completionToken, keywords, false, false);
+					// complete types as it is done for version < J17
+					findTypesAndPackages(this.completionToken, scope, false, false, new ObjectVector());
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java:3993]: as all this is under the flag `isSwitchEnumOrType` we shouldn't reach this line for string switches, but we do.

One indication for a true type switch would be that the resolvedType is not `final`, because then we don't have any subtypes to propose. Trying this in the test, reveals that `jclMin17.jar` contains a stub for `String` that does not declare the class as `final` (should be corrected anyway).

Alternatively, we could explicitly exclude from `isSwitchEnumOrType` if `resolvedType.id == TypeIds.T_javaLangString`.

**Review by stephan-herrmann — CHANGES_REQUESTED** [review]
I'm missing any tests with a switch over another reference type than `Object`.

E.g., I expect that for a switch over `Number` only types like `Integer`, `Double` ... are proposed. 

With this in mind the check for final-ness mentioned in another comment will become relevant.

**Comment by stephan-herrmann** [comment]
> E.g., I expect that for a switch over `Number` only types like `Integer`, `Double` ... are proposed.

For obvious reasons, searching for all subtypes is not feasible for certain types like `Object` or `Serializable`.

See org.eclipse.jdt.internal.ui.text.java.ContentAssistHistory.UNCACHEABLE for a list of types with huge sub hierarchy. I'm surprised I couldn't find anything in core's search engine to deal with those. Perhaps I didn't look in the right places?

**Comment by stephan-herrmann** [comment]
Another idea: for helping users new to this kind of switch, would it be prudent to insert not only a type, but also a name to which the value shall be bound? We could use existing heuristics for variable naming (`Integer i`, `Person person` ...).

**Comment by gayanper** [comment]
Thanks for the feedback @stephan-herrmann I will look into and improve the PR.

**Comment by gayanper** [comment]
@stephan-herrmann I did some changes as you suggest, Let me know if its go to merge

**Comment by gayanper** [comment]
> I'm missing any tests with a switch over another reference type than `Object`.
> 
> E.g., I expect that for a switch over `Number` only types like `Integer`, `Double` ... are proposed.
> 
> With this in mind the check for final-ness mentioned in another comment will become relevant.

The main purpose of the PR was to allow code completions after `case`, To improve completions for pattern matching we cloud report another issue and capture the expected behavior and work on it. @stephan-herrmann WDYT ?

**Comment by gayanper** [comment]
@stephan-herrmann The PR was made under the following assumptions

When we write a switch block with a int variable
```
switch(i) {
   case Const
}
```
completions behind `Const` will provide proposals where user can choose the class where they want to select a constant for example. if they don't do that, the compiler will issue a error.

Now I would see the same for a switch with a string variable like

```
switch(str) {
   case Const
}
```
But due to the code segment at https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java#L3967
which cause the scenario with str variable not to work. because the `else` condition at https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java#L3980 cause only to have switch block related keyword completions.

So the question is should we handle this scenario ? I feel we should since it is working for primitive and why not for string ? And the next question is why only limiting few keywords as completions when there is a non primitive resolved type in Java17 ? shouldn't the pattern specific errors must be handled by compiler ? And if we really want to support choosing right types in java17+ pattern matching, should we implement it based on completion relevance instead ?

**Comment by stephan-herrmann** [comment]
Thanks, @gayanper . I see now where you're coming from.
Indeed, the fix for [bug 575340](https://bugs.eclipse.org/bugs/show_bug.cgi?id=575340)  leaves some room for improvement, in that it addresses two keywords (`null` and `default`) which have different context conditions.

Indeed also, this PR improves the situation for one particular use case, but I'm afraid that this process will converge only very slowly, unless we ask the major questions first:
* what are the different kinds of switch statements?
* how can we detect them?
* what proposals should be offered for each kind?
  * keywords?
  * search based names? (what search strategy?)
  * more than names (entire patterns)?
* do different kinds of switch statements have commonalities that can be leveraged in the code (this question probably to be asked last)?

I predict that a condition like `isSwitchEnumOrType` will not stand the test of time, because it tries to generalize two situations that have very little in common.

I don't expect to solve all details at once, but tweaking corner cases within a sub-optimal overall design looks like an unprofitable investment of time.

**Comment by stephan-herrmann** [comment]
> And the next question is why only limiting few keywords as completions ...

what other keywords do you have in mind (in addition to `null` and `default`)?

**Comment by gayanper** [comment]
> > And the next question is why only limiting few keywords as completions ...
> 
> what other keywords do you have in mind (in addition to `null` and `default`)?

My question was not about keywords, But rather on other completions such as Type completions. Now the following is on Java11 project

image


same code with Java17 project

image

So as you can see regardless If its primitive or not (in this case String) we get Type completions in Java11 but not in Java17.

This is why I suggest we start with providing Type completion on Java17+ as well as we get in Java11. And then if needed we can fine grain the completions base on Patterns.

**Comment by stephan-herrmann** [comment]
Clickable link to the issue: #1301
